### PR TITLE
Converts remaining DAOs to jooq

### PR DESCRIPTION
**Pull request by alokmenghrajani**
I tried to split this into multiple smaller pieces but there is a transaction boundary in SecretDAO which makes it hard to mix jdbi and jooq.

### CODE REVIEW

File: server/src/main/java/keywhiz/ServiceModule.java [hunk 1]
```diff
@@ -181,12 +185,16 @@ public ServiceModule(KeywhizConfig config, Environment environment) {
   }
 
   @Provides @Singleton
-  @Readonly DBI readonlyDbi(DBIFactory factory, Environment environment, KeywhizConfig config,
-      MapArgumentFactory mapArgumentFactory, @Readonly ManagedDataSource dataSource)
+  @Readonly DBI readonlyDbi(KeywhizConfig config, MapArgumentFactory mapArgumentFactory,
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/ServiceModule.java]: DBI still needed?

File: server/src/main/java/keywhiz/service/daos/SecretContentDAO.java [hunk 2]
```diff
@@ -16,46 +16,99 @@
...
+
+import static keywhiz.jooq.tables.SecretsContent.SECRETS_CONTENT;
+
+/**
+ * Interacts with 'secrets_content' table and actions on {@link SecretContent} entities.
+ */
+public class SecretContentDAO {
+  private final DSLContext dslContext;
+  private final ObjectMapper
+      mapper = KeywhizService.customizeObjectMapper(Jackson.newObjectMapper());
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretContentDAO.java]: Inject ObjectMapper too

File: server/src/test/java/keywhiz/service/daos/SecretDAOTest.java [hunk 3]
```diff
@@ -93,8 +95,8 @@ public void setUp() throws Exception {
         .set(SECRETS.UPDATEDBY, secret2.series().getUpdatedBy())
         .execute();
 
-    testDBRule.jooqContext().insertInto(SECRETS_CONTENT)
-        .set(SECRETS_CONTENT.SECRETID, Math.toIntExact(secret2.series().getId()))
+    dslContext.insertInto(SECRETS_CONTENT)
+        .set(SECRETS_CONTENT.SECRETID, (int) secret2.series().getId())
```
Comment (sul3n3t) [inline server/src/test/java/keywhiz/service/daos/SecretDAOTest.java]: Why remove `Math.toIntExact()`?

File: server/src/main/java/keywhiz/ServiceModule.java (hunk 1 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/ServiceModule.java]: I have another cleanup PR after this one. You can diff my branch vs this
branch if you want to see what all will go away.

File: server/src/test/java/keywhiz/service/daos/SecretDAOTest.java (hunk 3 above)
Comment (alokmenghrajani) [inline server/src/test/java/keywhiz/service/daos/SecretDAOTest.java]: Bad merge, sorry.

File: server/src/main/java/keywhiz/service/daos/SecretContentDAO.java (hunk 2 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretContentDAO.java]: Fixed

File: server/src/main/java/keywhiz/service/daos/AclDAO.java [hunk 4]
```diff
@@ -52,14 +53,17 @@
   private GroupDAO groupDAO;
   private SecretContentDAO secretContentDAO;
   private SecretSeriesDAO secretSeriesDAO;
+  private ObjectMapper mapper;
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/AclDAO.java]: all the fields can be `final`

File: server/src/main/java/keywhiz/service/daos/SecretController.java [hunk 5]
```diff
@@ -73,6 +67,11 @@ public SecretController(SecretTransformer transformer, ContentCryptographer cryp
     return secretDAO.getSecretByNameAndVersion(name, version).map(transformer::transform);
   }
 
+  /** @return all existing secrets. */
+  public List<Secret> getSecrets() {
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretController.java]: I think @riyazdf deleted these three methods that are being added back here.

File: server/src/main/java/keywhiz/service/daos/SecretController.java [hunk 6]
```diff
@@ -179,9 +197,11 @@ public SecretBuilder withGenerationOptions(Map<String, String> generationOptions
      * @return an instance of the newly created secret.
      */
     public Secret build() {
-      try {
         secretDAO.createSecret(name, encryptedSecret, version, creator, metadata, description, type, generationOptions);
         return transformer.transform(secretDAO.getSecretByNameAndVersion(name, version).get());
+/*
+        // TODO: I don't know if propagating the DataAccessException is the right thing to do.
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretController.java]: Are you considering whether it's the correct type or if it should be caught here?

Comments/catches below can be deleted.

File: server/src/main/java/keywhiz/service/daos/SecretDAO.java [hunk 7]
```diff
@@ -38,46 +37,60 @@
  * Does not map to a table itself, but utilizes both {@link SecretSeriesDAO} and
  * {@link SecretContentDAO} to provide a more usable API.
  */
-public abstract class SecretDAO implements Transactional<SecretDAO> {
-  @CreateSqlObject protected abstract SecretContentDAO createSecretContentDAO();
-  @CreateSqlObject protected abstract SecretSeriesDAO createSecretSeriesDAO();
+public class SecretDAO {
+  private final DSLContext dslContext;
+  private SecretContentDAO secretContentDAO;
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretDAO.java]: `final` on these two fields

File: server/src/main/java/keywhiz/service/daos/SecretDAO.java [hunk 8]
```diff
@@ -85,36 +98,42 @@ public long createSecret(String name, String encryptedSecret, String version,
...
-
-    Optional<SecretContent> content =
-        createSecretContentDAO().getSecretContentBySecretIdAndVersion(secretId, version);
-    if (!content.isPresent()) {
-      return Optional.empty();
-    }
-
-    return Optional.of(SecretSeriesAndContent.of(series.get(), content.get()));
+
+    // Cast to fix issue with mvn + java8 (not showing up in Intellij).
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretDAO.java]: Comment can go

File: server/src/main/java/keywhiz/service/daos/SecretDAO.java [hunk 9]
```diff
@@ -85,36 +98,42 @@ public long createSecret(String name, String encryptedSecret, String version,
...
-  @Transaction
   public ImmutableList<String> getVersionsForSecretName(String name) {
     checkNotNull(name);
-    Optional<SecretSeries> series = createSecretSeriesDAO().getSecretSeriesByName(name);
-    if (!series.isPresent()) {
-      return ImmutableList.of();
-    }
 
-    return createSecretContentDAO().getVersionFromSecretId(series.get().getId());
+    // Cast to fix issue with mvn + java8 (not showing up in Intellij).
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretDAO.java]: Ditto, comment can go

File: server/src/main/java/keywhiz/service/daos/SecretDAO.java [hunk 10]
```diff
@@ -122,38 +141,40 @@ public long createSecret(String name, String encryptedSecret, String version,
...
 
-    Optional<SecretSeries> secretSeries = createSecretSeriesDAO().getSecretSeriesByName(name);
-    if (!secretSeries.isPresent()) {
-      return Optional.empty();
-    }
-
-    SecretContentDAO secretContentDAO = createSecretContentDAO();
-    Optional<SecretContent> secretContent =
-        secretContentDAO.getSecretContentBySecretIdAndVersion(secretSeries.get().getId(), version);
+    // Cast to fix issue with mvn + java8 (not showing up in Intellij).
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretDAO.java]: Ditto, comment can go

File: server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java [hunk 11]
```diff
@@ -16,46 +16,89 @@
...
-  void deleteSecretSeriesByName(@Bind("name") String name);
-
-  @SqlUpdate("DELETE FROM secrets WHERE id = :id")
-  void deleteSecretSeriesById(@Bind("id") long id);
+import keywhiz.jooq.tables.records.SecretsRecord;
+import org.jooq.DSLContext;
+
+import static keywhiz.jooq.tables.Secrets.SECRETS;
+
+public class SecretSeriesDAO {
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java]: bring over javadoc from prev version

File: server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java [hunk 12]
```diff
@@ -16,46 +16,89 @@
...
+  private final DSLContext dslContext;
+  private final ObjectMapper
+      mapper = KeywhizService.customizeObjectMapper(Jackson.newObjectMapper());
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java]: inject the objectmapper

File: server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java [hunk 13]
```diff
@@ -16,46 +16,89 @@
...
+  @Inject
+  public SecretSeriesDAO(DSLContext dslContext) {
+    this.dslContext = dslContext;
+  }
+
+  long createSecretSeries(String name, String creator, String description, @Nullable String type,
+      @Nullable Map<String, String> generationOptions) {
+    SecretsRecord r =  dslContext.newRecord(SECRETS);
+
+    r.setName(name);;
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java]: nit: extra semicolon

File: server/src/main/java/keywhiz/service/daos/SecretContentMapper.java [hunk 14]
```diff
@@ -19,54 +19,47 @@
...
+        r.getEncryptedContent(),
+        r.getVersion(),
+        r.getCreatedat(),
+        r.getCreatedby(),
+        r.getUpdatedat(),
+        r.getUpdatedby(),
+        tryToReadMapFromMetadata(r));
+  }
+
+  private ImmutableMap<String, String> tryToReadMapFromMetadata(SecretsContentRecord r) {
```
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretContentMapper.java]: nit: implemented differently than SecretSeriesMapper

File: server/src/main/java/keywhiz/service/daos/AclDAO.java (hunk 4 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/AclDAO.java]: Fixed

File: server/src/main/java/keywhiz/service/daos/SecretContentMapper.java (hunk 14 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretContentMapper.java]: Yes. This gets cleaned up when I convert everything to FooRecord (https://github.com/square/keywhiz/commit/65e1ca505e067efc4da7d6b693e7a015b95863de)

File: server/src/main/java/keywhiz/service/daos/SecretController.java (hunk 5 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretController.java]: Removed

File: server/src/main/java/keywhiz/service/daos/SecretController.java (hunk 6 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretController.java]: Two questions:
1. should we catch DataAccessException (which is a jooq thing) and throw something more generic?
2. does DataAccessException leak any information we don't want to leak (sensitive data from raw SQL)?

File: server/src/main/java/keywhiz/service/daos/SecretDAO.java (hunk 7 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretDAO.java]: Fixed

File: server/src/main/java/keywhiz/service/daos/SecretDAO.java (hunk 8 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretDAO.java]: Fixed

File: server/src/main/java/keywhiz/service/daos/SecretDAO.java (hunk 9 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretDAO.java]: Fixed

File: server/src/main/java/keywhiz/service/daos/SecretDAO.java (hunk 10 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretDAO.java]: Fixed

File: server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java (hunk 11 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java]: Fixed

File: server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java (hunk 12 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java]: Fixed

File: server/src/main/java/keywhiz/service/daos/AclDAO.java (hunk 4 above)
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/AclDAO.java]: Not sure what the JVM infers. Causes a compilation error if you forget to initialize it and makes it clearer to readers which fields are expected to be modifiable.

File: server/src/main/java/keywhiz/service/daos/SecretController.java (hunk 6 above)
Comment (sul3n3t) [inline server/src/main/java/keywhiz/service/daos/SecretController.java]: 1. I'm okay with DataAccessException for now. It is a namespaced RuntimeException. We could unwrap/rewrap the cause, but I can't think of a reason right now. You?
2. If there is a leak, it'd be from exceptions nested in DataAccessException. That'd be hard to test exhaustively, generating SQL errors like constraint violations would be a good sanity check.

File: server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java (hunk 13 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretSeriesDAO.java]: fixed

File: server/src/main/java/keywhiz/service/daos/AclDAO.java (hunk 4 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/AclDAO.java]: Good point about forgetting to initialize.

File: server/src/main/java/keywhiz/service/daos/SecretController.java (hunk 6 above)
Comment (alokmenghrajani) [inline server/src/main/java/keywhiz/service/daos/SecretController.java]: Let's go with DataAccessException for now. I'll sanity check that nothing leaks.

**Comment by sul3n3t** [comment]
:+1:

**Pull request merged**